### PR TITLE
SqlObject implicit @Bind by parameter name

### PR DESCRIPTION
**Pull request by stevenschlansker**
We added support for `@Bind` recognizing parameter names, now let's go even further and ditch the annotation if you don't actually need it.

Open question: should we just get rid of the default "name by parameter index" entirely?  This wouldn't prevent people from using positional parameters in fluent queries, but it would mean
```
@SqlUpdate("insert into table values(:0)")
void doSomething(int param);
```
would need to be rewritten to
```
@SqlUpdate("insert into table values(:param)")
void doSomething(int param);
```

The only reason I can see this being used would be if your code does not have parameter names enabled (which is a javac flag IIRC) but then you might as well just `@Bind("param")` instead... Thoughts?

### CODE REVIEW

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultObjectBinder.java [hunk 1]
```diff
@@ -19,20 +19,36 @@
...
+                value = Integer.toString(paramIndex);
+            } else {
                 throw new UnsupportedOperationException("A parameter was annotated with @Bind "
                         + "but no name was specified, and parameter name data is not present "
                         + "in the class file.  " + param.getDeclaringExecutable() + " :: " + param);
             }
-            value = param.getName();
         }
 
+        q.bind(paramIndex, arg);
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultObjectBinder.java]: Should this be called if this `DefaultObjectBinder` is created with the no-arg constructor?

**Comment by qualidafial** [comment]
Do we really want to drop positional parameter support? We're sure there aren't enough users that use it to justify keeping it?

I'm also not sure I like falling back to positional parameters behavior depending on the user's compiler flags. IMO it should only work (or not) depending on compiler flags, not change behavior.

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultObjectBinder.java (hunk 1 above)
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultObjectBinder.java]: The more I look at this, the more I think we don't even want this class doing anything positional.

**Comment by stevenschlansker** [comment]
To clarify, I wasn't suggesting removing positional parameter support.  There was a bizarre mode where `@Bind` would create a *named* parameter with the stringified position index as a name.  So it's a named positional parameter.  Confusing!

Now that's gone.  I simplified the rules a lot:

It's always bound as a positional parameter, which will only be used if the name doesn't match.
If you omit the name, you *must* have the parameter name available, otherwise it's an error.

Think that's reasonable?

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/Bind.java [hunk 2]
```diff
@@ -19,7 +19,11 @@
 import java.lang.annotation.Target;
 
 /**
- * A binding annotation
+ * A binding annotation.  Will bind both as the {@code value}
+ * of the annotation as well as the parameter index.  If the
+ * {@code value} is not specified, it will default to using the parameter
+ * name.  It is an error to not specify the name when there is no
+ * parameter naming information in your class files.
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/Bind.java]: :+1:

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/BinderFactory.java [hunk 3]
```diff
@@ -19,12 +19,12 @@
  * Factor for building {@link Binder} instances. This interface is used by
  * the {@link BindingAnnotation}
  */
-public interface BinderFactory
+public interface BinderFactory<T extends Annotation, O>
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/BinderFactory.java]: Should add documentation for `T` and `O` generic parameters.

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultObjectBinder.java [hunk 4]
```diff
@@ -19,20 +19,35 @@
 
 class DefaultObjectBinder implements Binder<Bind, Object>
 {
+    private final int paramIndex;
+
+    DefaultObjectBinder()
+    {
+        this(-1);
+    }
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultObjectBinder.java]: Is this no-arg constructor necessary?

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/CustomizingStatementHandler.java [hunk 5]
```diff
@@ -85,8 +85,8 @@
                     // we have a binder
                     BindingAnnotation ba = annotation.annotationType().getAnnotation(BindingAnnotation.class);
                     try {
-                        BinderFactory fact = ba.value().newInstance();
-                        binders.add(new Bindifier(this.method, annotation, param_idx, (Binder<Annotation, Object>)fact.build(annotation)));
+                        BinderFactory<Annotation, Object> fact = (BinderFactory<Annotation, Object>) ba.value().newInstance();
+                        binders.add(new Bindifier<Annotation>(this.method, annotation, param_idx, fact.build(annotation)));
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/CustomizingStatementHandler.java]: The presence of a generic cast here suggests that `Bindifier` should expect a `BinderFactory<AnnoType extends Annotation, ?>`. Bindifier isn't using the second parameter.

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/BinderFactory.java (hunk 3 above)
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/BinderFactory.java]: I looks like the `O` parameter is unused.

**Comment by qualidafial** [comment]
Consider comments, then :shipit: when ready.

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultObjectBinder.java (hunk 4 above)
Comment (stevenschlansker) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultObjectBinder.java]: yeah, if you have `@Bind(binder=DefaultObjectBinder.class)`, it calls `DefaultObjectBinder.class.newInstance()`

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/CustomizingStatementHandler.java (hunk 5 above)
Comment (stevenschlansker) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/CustomizingStatementHandler.java]: added a TODO to clean up the generics

**Comment by stevenschlansker** [comment]
I think we should rethink how binding happens entirely.  I've dropped a note to that effect on the CHANGES_FOR_3.0 document and we'll revisit it later.  I'm going to merge this partially as a WIP.

**Pull request merged**